### ISSUE TITLE
Better project options

### ISSUE THREAD

**Issue by jpakkane** [mesonbuild/meson#707]
The current project option setup is simple. All options must be specified at the top level `meson_options.txt` file. We should instead be able to provide roughly similar level of options as kconfig. These include:

- some options that are not shown, for example if you have `win`, `linux` and `osx` subdirectories, each with options, only options from one of them should be added to the set of all options and shown to the user
- arranging options in a tree
- something like `make allmodconfig` seems useful but hard to do reliably with all of the above

KConfig is nice but has severe drawbacks such as being able to set configurations that are broken or inconsistent. We do not want that.

**Comment by marc-h38** [comment]
Obsolete?

**Comment by bonzini** [comment]
Maybe the option interpreter should have access to a bit more information such as host_machine/build_machine/target_machine and `if` or even `foreach` statements.

**Comment by eli-schwartz** [comment]
That seems like a very bad solution to the problem.

Meson has gotten a lot of mileage to date using an options definition schema which is trivially parseable via eyeball, and which guarantees that all options are always defined / available via `get_option()` and the available list of options cannot change without editing the options file itself. I don't see a compelling need to change this.

Adding `if` support definitely seems like a step too far:

- Arranging options in a tree is an interesting idea but entirely orthogonal to having `if` support.

- What has been periodically floated as a good idea is having options with different default values depending on system type, which could be done by adding a new kwarg to option() without granting it full flow control. I could also hear having options that are only valid for a specific system type, but if we did support such a thing I would *really* want users to get a proper error message if they tried to use it for the wrong system type ("ERROR: option windows-foo can only be selected on Windows"), which means defining the function all the time but having a kwarg (again) to control availability.



foreach doesn't seem useful to anything mentioned here, if it were useful for anything it would be useful for creating a large number of options that all share a particular bit of information. It's just code generation implemented as DRY, not an actual decision-making process. I don't think it is particularly useful *in general* either, since options are already quite lightweight and it doesn't seem plausible that it's going to be better than spelling them out.

**Comment by dcbaker** [comment]
I agree with @eli-schwartz on this. the three things I've wished for are:
 - different defaults depending on the platform. For me, this can be handled in the meson.build pretty trivially with `feature` and the various `enable_*`/`disable_*` methods combined with the various bool methods `allowed`, `enabled`, `disabled`. So I could consider this to be a lower priority for me
 - options that have per-machine values, or, conversely, are explicitly one machine. The problem with this though is `meson.build.` enforcement of using `build` options on `host` targets. In other-words, having `-Dbuild.user_option=...` and `-Duser_optiion=...` like `build.c_args` and `c_args`
 - options that are only used on specific platforms. I think though from a portability point of view I'd want these to just be hidden in printed messages like `meson configure $builddir`. Seems difficult to deal with
